cirq/optimizers/controlled_gate_decomposition.py

```diff
@@ -0,0 +1,254 @@
+        ]
+        gates2 = _ccnot_congruent(controls[0], controls[1], free_qubits[0])
+        gates3 = _flatten(gates1) + gates2 + _flatten(gates1[::-1])
+        first_ccnot = ops.CCNOT(controls[m - 1], free_qubits[m - 3], target)
+        return [first_ccnot, *gates3, first_ccnot, *gates3]
+    elif len(free_qubits) >= 1:
+        # See [1], Lemma 7.3.
+        m1 = n // 2
+        free1 = controls[m1:] + [target] + free_qubits[1:]
+        ctrl1 = controls[:m1]
+        part1 = decompose_multi_controlled_x(ctrl1, free_qubits[0], free1)
+        free2 = controls[:m1] + free_qubits[1:]
+        ctrl2 = controls[m1:] + [free_qubits[0]]
+        part2 = decompose_multi_controlled_x(ctrl2, target, free2)
+        return [*part1, *part2, *part1, *part2]
+    else:
+        # No free qubits - must use general algorithm.
+        # This will never happen if called from main algorithm and is added
+        # only for completeness.
+        return decompose_multi_controlled_unitary(unitary(ops.X), controls,
+                                                  target)
+
+
+def _decompose_su(matrix: np.ndarray, controls: List['cirq.Qid'],
+                  target: 'cirq.Qid') -> List['cirq.Operation']:
+    """Decomposes controlled special unitary gate into elementary gates.
+
+    Result has O(len(controls)) operations.
+    See [1], lemma 7.9.
+    """
+    assert matrix.shape == (2, 2)
+    assert is_special_unitary(matrix)
+    assert len(controls) >= 1
+
+    a, b, c, _ = _decompose_abc(matrix)
+
+    cnots = decompose_multi_controlled_x(controls[:-1], target, [controls[-1]])
+    return [
+        *_decompose_single_ctrl(c, controls[-1], target), *cnots,
+        *_decompose_single_ctrl(b, controls[-1], target), *cnots,
+        *_decompose_single_ctrl(a, controls[-1], target)
+    ]
+
+
+def _decompose_recursive(matrix: np.ndarray, power: float,
+                         controls: List['cirq.Qid'], target: 'cirq.Qid',
+                         free_qubits: List['cirq.Qid']
+                        ) -> List['cirq.Operation']:
+    """Decomposes controlled unitary gate into elementary gates.
+
+    Result has O(len(controls)^2) operations.
+    See [1], lemma 7.5.
+    """
+    if len(controls) == 1:
+        return _decompose_single_ctrl(_unitary_power(matrix, power),
+                                      controls[0], target)
+
+    cnots = decompose_multi_controlled_x(controls[:-1], controls[-1],
+                                         free_qubits + [target])
+    return [
+        *_decompose_single_ctrl(_unitary_power(matrix, 0.5 * power),
+                                controls[-1], target), *cnots,
+        *_decompose_single_ctrl(_unitary_power(matrix, -0.5 * power),
+                                controls[-1], target), *cnots,
+        *_decompose_recursive(matrix, 0.5 * power, controls[:-1], target,
+                              [controls[-1]] + free_qubits)
+    ]
+
+
+def decompose_multi_controlled_unitary(matrix: np.ndarray,
```

Comment:
Rename to `decompose_multi_controlled_rotation` to avoid ambiguity between single-qubit and multi-qubit unitaries.

Reply:
Done.